File: test/src/Channel.cc
// Input Channel
//
// Created by Samvel Khalatyan, Dec 2, 2011
// Copyright 2011, All rights reserved

#include <ostream>

#include "interface/Channel.h"

using namespace std;

Channel::operator string() const
{
    switch(type())
    {
        case DATA: return "CMS Data 2011";
        case QCD: return "QCD data-driven";
        case TTBAR: return "t#bar{t}";
        case WJETS: return "W#rightarrowl#nu";
        case ZJETS: return "Z/#gamma*#rightarrowl^{+}l^{-}";
        case STOP: return "Single-Top";
        case ZPRIME1000: return "Z' 1 #pm .01 TeV/c^{2}";
        case ZPRIME1500: return "Z' 1.5 #pm .015 TeV/c^{2}";
        case ZPRIME2000: return "Z' 2 #pm .02 TeV/c^{2}";

        default: return "unknown";
    }
}

string Channel::repr() const
{
    switch(type())
    {
        case DATA: return "data_2011";
        case QCD: return "qcd";
        case TTBAR: return "ttbar";
        case WJETS: return "wjets";
        case ZJETS: return "zjets";
        case STOP: return "stop";
        case ZPRIME1000: return "zprime_1000";
        case ZPRIME1500: return "zprime_1500";
        case ZPRIME2000: return "zprime_2000";

        default: return "unknown";
    }
}

Channel &Channel::operator ++()
{
    if (ZPRIME2000 != type());
        _type = Type(static_cast<int>(type()) + 1);

    return *this;
}

Channel &Channel::operator --()
{
    if (DATA != type());
        _type = Type(static_cast<int>(type()) - 1);

    return *this;
}



// Helpers
//
bool operator <(const Channel &c1, const Channel &c2)
{
    return c1.type() < c2.type();
}

bool operator >(const Channel &c1, const Channel &c2)
{
    return c1.type() > c2.type();
}

bool operator <=(const Channel &c1, const Channel &c2)
{
    return c1.type() <= c2.type();
}

bool operator >=(const Channel &c1, const Channel &c2)
{
    return c1.type() >= c2.type();
}

bool operator ==(const Channel &c1, const Channel &c2)
{
    return c1.type() == c2.type();
}

bool operator !=(const Channel &c1, const Channel &c2)
{
    return c1.type() != c2.type();
}

ostream &operator <<(ostream &out, const Channel &c)
{
    return out << static_cast<string>(c);
}

File: python/template/channel.py
#!/usr/bin/env python

'''
Created by Samvel Khalatyan, Feb 22, 2012
Copyright 2011, All rights reserved
'''

from base_type import BaseType

class ChannelType(BaseType):
    '''
    Specify channel types for plots. These are merged input types in some
    cases, for example: stop_s, stop_t, stop_tw, etc. are merged into stop
    (single top)
    '''

    # Generate allowed channel types from input types and merge single tops
    # into one channel
    channel_types = set([
            "ttbar",
            "zjets",
            "wjets",
            "stop",
            "data",
            "zprime_m1000_w10",
            "zprime_m1500_w15",
            "zprime_m2000_w20",
            "zprime_m3000_w30",
            "zprime_m4000_w40"
            ])

    def __init__(self, channel_type):
        BaseType.__init__(self, channel_type, "channel_type")

    def __contains__(self, value):
        return (value in self.channel_types or
                BaseType.__contains__(self, value))

if "__main__" == __name__:
    import unittest

    class TestChannelType(unittest.TestCase):
        def test_ttbar(self):
            input_type = ChannelType("ttbar")
            self.assertEqual(input_type.type, "ttbar")

        def test_zjets(self):
            input_type = ChannelType("zjets")
            self.assertEqual(input_type.type, "zjets")

        def test_wjets(self):
            input_type = ChannelType("wjets")
            self.assertEqual(input_type.type, "wjets")

        def test_stop(self):
            input_type = ChannelType("stop")
            self.assertEqual(input_type.type, "stop")

        def test_data(self):
            input_type = ChannelType("data")
            self.assertEqual(input_type.type, "data")

        def test_zprime_m1000_w10(self):
            input_type = ChannelType("zprime_m1000_w10")
            self.assertEqual(input_type.type, "zprime_m1000_w10")

        def test_zprime_m1500_w15(self):
            input_type = ChannelType("zprime_m1500_w15")
            self.assertEqual(input_type.type, "zprime_m1500_w15")

        def test_zprime_m2000_w20(self):
            input_type = ChannelType("zprime_m2000_w20")
            self.assertEqual(input_type.type, "zprime_m2000_w20")

        def test_zprime_m3000_w30(self):
            input_type = ChannelType("zprime_m3000_w30")
            self.assertEqual(input_type.type, "zprime_m3000_w30")

        def test_zprime_m4000_w40(self):
            input_type = ChannelType("zprime_m4000_w40")
            self.assertEqual(input_type.type, "zprime_m4000_w40")

        def test_qcd(self):
            self.assertRaises(AttributeError, ChannelType, "qcd")

    unittest.main()
